### ISSUE TITLE
Add basic serving support for PYPTO

### ISSUE THREAD

**Issue by superxf** [hw-native-sys/pypto-serving#5]
# PyPTO Serving End-to-End Flow Description

---

### 1.1 Goals

- Provide HTTP APIs similar to the **OpenAI API style** (`/v1/completions`, `/v1/chat/completions`, etc.).
- The main process handles **scheduling, batch decision-making, and HTTP**, while worker subprocesses exclusively own **NPU + model forward pass**, avoiding cross-process large Tensor transfers and device thread affinity issues.
- Support **continuous batching**, **chunked prefill**, **prefix caching**, and **preemption**.

### 1.2 Dual-Process Architecture (default NPU path)

| Role | Responsibilities |
|------|-----------------|
| **Main Process** | FastAPI service, `AsyncLLMEngine`, `Scheduler`, `BlockPool`, Tokenizer (vocabulary encoding/decoding only), communicates with Worker via queues |
| **Worker Subprocess** | `WorkerProcess`: loads model and executor, `KvCacheManager`, Prefill/Decode kernels, `Sampler` |

Queue direction:

```text
Main Process --[input_queue: WorkerCommand]--> Worker
Worker --[output_queue: StepOutput]--> Main Process
```

#### Diagram A: Uvicorn startup/shutdown and collaboration with Engine and Worker

```mermaid
sequenceDiagram
    autonumber
    participant UV as Uvicorn
    participant APP as FastAPI
    participant ENG as AsyncLLMEngine (Main Process)
    participant WK as Worker Subprocess

    rect rgb(245, 250, 255)
        Note over UV,WK: Startup (lifespan)
        UV->>APP: lifespan startup
        APP->>ENG: await start()
        ENG->>WK: spawn_worker / Process.start
        WK->>WK: init_device_and_model()
        WK-->>ENG: ready_event.set()
        ENG->>ENG: create_task(_engine_loop)
        ENG-->>APP: start returns
        APP-->>UV: startup complete
    end

    rect rgb(255, 250, 245)
        Note over UV,WK: Shutdown (lifespan)
        UV->>APP: lifespan shutdown
        APP->>ENG: await stop()
        ENG->>ENG: _running=False, await _loop_task
        ENG->>WK: WorkerCommand(shutdown)
        WK->>WK: busy_loop exits
        ENG->>WK: join / terminate
        ENG-->>APP: stop returns
    end
```

#### Diagram B: Concurrent relationship of one HTTP generation and "schedule → Worker → enqueue response"

In a single request, the **routing coroutine** awaits `ctx.queue.get()` in `add_request`; the **`_engine_loop`** coroutine runs independently, handling `schedule`, exchanging one step of results with the Worker, and writing `TokenOutput` to that request's `ctx.queue`. Both share `AsyncLLMEngine` and `Scheduler`, with asyncio interleaving execution.

```mermaid
sequenceDiagram
    autonumber
    participant C as Client
    participant S as ServingServer
    participant E as AsyncLLMEngine
    participant W as Worker Subprocess

    C->>S: POST /v1/completions (or chat)
    S->>E: add_request (encode, scheduler.add_request)

    loop until TokenOutput.finished
        par Coroutine handling this HTTP
            E->>E: await ctx.queue.get (suspended until output)
        and Background _engine_loop
            E->>E: scheduler.schedule()
            E->>W: WorkerCommand(step) + input_queue
            W->>W: free completed req (if any) / prefill / decode / sample
            W-->>E: StepOutput → output_queue
            E->>E: update_from_output, decode, stop detection
            E->>E: ctx.queue.put_nowait(TokenOutput)
        end
        E-->>S: yield TokenOutput
        S-->>C: JSON or SSE data frame
    end
```

#### Diagram C: Difference between non-streaming and streaming on the ServingServer side (engine side is the same)

```mermaid
sequenceDiagram
    participant C as Client
    participant S as ServingServer
    participant E as AsyncLLMEngine

    alt stream=false
        C->>S: POST ... stream=false
        loop async for until finished
            S->>E: add_request
            E-->>S: TokenOutput
        end
        S->>C: Single JSON (full text + finish_reason)
    else stream=true
        C->>S: POST ... stream=true
        S->>C: StreamingResponse starts
        loop async for until finished
            S->>E: add_request
            E-->>S: TokenOutput
            S->>C: SSE: delta = incremental text relative to previous frame
        end
        S->>C: data: [DONE]
    end
```

---

### Step 1: Parse CLI

Entry point: `llm/cli/main.py` `build_parser()` defines `pypto-serving`.

- Must specify **exactly one** mode: `--prompt`, `--interactive`, or **`--serve`** (mutually exclusive).
- Serving-related parameters include: `--config`, `--host`, `--port`, `--max-num-running-reqs`, `--max-num-scheduled-tokens`, `--long-prefill-token-threshold`, etc.

### Step 2: Load JSON config `load_serving_config`

- Reads sections like `model`, `runtime`, `generation`, `npu`, constructs CLI-side `ServingConfig` (different name from `async_engine.ServingConfig`, do not confuse).
- **Note**: When using `--serve`, the main process does **not** call `create_engine` / `init_engine` to load the full model weights locally (see `main()` branch), to reduce main process burden and avoid duplicate GPU usage with Worker.

### Step 3: `run_serve()` assembles async engine

1. **Tokenizer**: Main process uses `TransformersTokenizerAdapter.from_pretrained(model_dir)` to load only the tokenizer, for `encode` / `decode`.
2. **`WorkerConfig`**: Write `model_id`, `model_dir`, `platform`, `device_id`, `runtime_config`, `executor_cls` (NPU is `PyptoQwen14BExecutor`, CPU is `ModelExecutor`), `executor_kwargs` (such as `l3_mode`, etc.).
3. **`AsyncServingConfig`** (`ServingConfig` in `llm/core/async_engine.py`): Maps CLI concurrency and scheduling limits, for example:
   - `max_num_running_reqs`: Upper limit for concurrent running requests;
   - `max_num_scheduled_tokens`: Maximum token budget per `schedule()` call (dynamic batch size upper bound);
   - `long_prefill_token_threshold`: Maximum tokens for a single prefill of long prompts (chunked prefill);
   - `max_seq_len`, `block_size` (usually consistent with `runtime.page_size`).
4. Construct **`AsyncLLMEngine`**, and **`create_serving_app(async_engine, model_id)`** to get FastAPI `app`.

### Step 4: FastAPI lifecycle hooks

- **`startup`**: `await async_engine.start()`
  - If not `in_process`: `spawn_worker()` starts subprocess, inside subprocess `_worker_entry` → `init_device_and_model()` → `ready_event.set()` → `busy_loop()`; main process `await asyncio.to_thread(ready_event.wait, timeout=600)` waits for model readiness.
  - Then `asyncio.create_task(self._engine_loop())` starts engine main loop.
- **`shutdown`**: `await async_engine.stop()`: stops loop, sends `shutdown` command, joins/terminates subprocess.

### Step 5: `uvicorn.run`

Binds `--host` / `--port`, starts accepting HTTP requests.

---

## 3. HTTP Request Handling (OpenAI-style API)

Implementation: `ServingServer` in `llm/core/server.py`.

### Step 1: Routes and model metadata

| Method | Path | Purpose |
|--------|------|---------|
| GET | `/health` | Health check |
| GET | `/v1/models` | Returns current `model_id` |
| POST | `/v1/completions` | Text completion |
| POST | `/v1/chat/completions` | Chat completion (assembles prompt first) |

### Step 2: Chat template `_apply_chat_template`

Current implementation is a **simple placeholder template** (`<|system|>` / `<|user|>` / `<|assistant|>` concatenation), **not** HuggingFace `tokenizer.apply_chat_template`. Production environments can replace with model-consistent template.

### Step 3: Construct `GenerateConfig`

Maps HTTP fields to `GenerateConfig`: `max_new_tokens` ← `max_tokens`, `temperature`, `top_p`, `top_k`, `stop`, `stream`.

### Step 4: Non-streaming response

- **`async for`** on `self.engine.add_request(request_id, prompt, config)`, accumulates the last non-empty `output.text` as full text, at the end takes `finish_reason` and maps to OpenAI style (`_map_finish_reason`).
- Returns JSON.

### Step 5: Streaming response (SSE)

- Returns `StreamingResponse`, `media_type="text/event-stream"`.
- Each time a `TokenOutput` is received from the engine, uses **incremental delta relative to previous frame** `delta = output.text[len(prev_text):]` as chunk content (avoids re-sending entire text).
- At the end, sends `data: [DONE]\n\n`.

---

## 4. Single request enters engine: `AsyncLLMEngine.add_request`

Implementation: `llm/core/async_engine.py`.

1. **Generate `request_id`**: Passed from routing layer (e.g., `cmpl-xxxx` / `chatcmpl-xxxx`).
2. **Tokenization**: `prompt_token_ids = tokenizer.encode(prompt)`; if empty and `bos_token_id` is configured, insert BOS; still empty then error.
3. **Construct `Request`**: Write `max_new_tokens`, `stop_strings`, `eos_token_id`, sampling parameters, etc.
4. **Register context**: `_RequestContext(request, asyncio.Queue)` stored in `_request_contexts[request_id]`.
5. **`scheduler.add_request(request)`**: Request enters waiting queue and participates in subsequent continuous batch scheduling.
6. **Async consumption**: `while True: output = await ctx.queue.get(); yield output; if output.finished: break`.
7. **`finally`**: Remove from `_request_contexts` (request ends or generator exits with exception).

---

## 5. Engine main loop: `_engine_loop` (each iteration)

Main process loops while `_running` is true.

### Step 1: Idle wait

- If `not scheduler.has_work()`: `await asyncio.sleep(engine_loop_interval)` (default ~1ms), avoids busy-spinning on CPU.

### Step 2: Schedule `scheduler.schedule()`

- If `scheduler_output.is_empty`: also briefly sleep.
- Otherwise, gets list of **`ScheduledRequest`** to execute this step (may mix prefill and decode) and token statistics.

### Step 3: Send Worker command

- Packs **`finished_request_ids`** (`_pending_free_ids`) from previous step with current **`scheduler_output`** into:

  `WorkerCommand(type="step", scheduler_output=..., finished_request_ids=...)`

- `input_queue.put(...)`.

### Step 4: Blocking wait for Worker result

- `step_output = await asyncio.to_thread(output_queue.get, timeout=300)`
  Uses thread wait for multiprocessing queue in async context to avoid blocking event loop for too long while still waiting for NPU step completion.

### Step 5: Error handling

- If `step_output.error`: logs, `_handle_step_error` calls `abort_request` for this batch and pushes `finished=True, finish_reason="error"` to each `ctx.queue`.

### Step 6: Process normal results `_process_step_output`

See next section.

---

## 6. Scheduler `Scheduler.schedule()` two phases (dynamic batching)

Implementation: `llm/core/scheduler.py`.

**Global constraint**: `token_budget = max_num_scheduled_tokens`.

### Phase A: Requests already in `running` (decode or chunked prefill continuation)

For each `running` request:

1. Calculate `num_new_tokens` needed this step (limited by `long_prefill_token_threshold` and remaining `token_budget`).
2. Calculate **`_blocks_needed`** based on `block_pool`, attempt **`_try_allocate_blocks`**.
3. If blocks insufficient: attempt **`_preempt_lowest_priority`** (preempts requests with latest arrival time first, resets to `PREEMPTED` and returns to front of `waiting` queue), then retry allocation.
4. On success, append `ScheduledRequest(request, num_new_tokens, is_prefill=request.is_prefill)`, and deduct `token_budget`.

### Phase B: New requests in `waiting` queue (new prefill)

When `token_budget` still has capacity and `len(running) < max_num_running_reqs`:

1. Take request from left side of `waiting`.
2. **Prefix caching**: If `block_pool.get_computed_blocks(prompt)` hits, set `cached_block_ids` and `num_computed_tokens` (skips length corresponding to cached prefix blocks).
3. Also subject to chunk threshold and block allocation constraints; on success, move to `running` and add to `scheduled_requests` this step.

**Key point**: Single-step `SchedulerOutput` can contain **multiple requests' prefill fragments + multiple decodes**, implementing **continuous batching**; long prefill split across multiple steps is **chunked prefill**.

---

## 7. Worker-side step execution: `WorkerProcess.busy_loop` / `_execute_step`

Implementation: `llm/core/worker.py`.

### Step 1: Process command

- `shutdown`: exits loop.
- `step`: If `finished_request_ids` present, first **`free_allocation`** (releases pages for that request in `KvCacheManager`).

### Step 2: Split prefill / decode

- `prefill_requests`: where `sr.is_prefill` is true.
- `decode_requests`: otherwise.

### Step 3: Execute inside `executor.session()`

1. **If prefill present**: `_batch_prefill`
   - Batch tensors, lookup embedding, `executor.run_prefill`; for requests **whose entire prompt was just computed this step**, sample the first generated token from the prefill's last layer logits and write to `new_tokens[request_id]`.
2. **If decode present**: `_batch_decode`
   - Take last token from previous step (or prompt's last token), expand KV slot, `executor.run_decode`, sample for each sequence's logits and write to `new_tokens`.

### Step 4: Return `StepOutput`

- `StepOutput(new_tokens={req_id: token_id, ...})`; if exception, `error` field is non-empty.

---

### Step 1: `scheduler.update_from_output(scheduler_output, step_output.new_tokens)`

- For each `ScheduledRequest` this step: increment `num_computed_tokens += num_new_tokens`, and **`_cache_completed_blocks`** updates prefix cache metadata.
- **Prefill incomplete**: no new token output (continue).
- **Prefill complete or decode**: take token from `new_token_ids`, append to `request.output_token_ids`, and produce `RequestOutput`.
- Then **`_check_finish`**: EOS, `max_new_tokens` limit reached marks `FINISHED_EOS` / `FINISHED_LENGTH`, remove from `running` and **`_free_request_blocks`**.

### Step 2: Main process `_process_step_output`

1. Use **`tokenizer.decode(request.output_token_ids)`** again to get current **complete generated text** (used for HTTP layer full or delta computation).
2. **`stop_strings`**: Scheduler doesn't handle string stops; if decoded text ends with some `stop`, force end, `FINISHED_STOP`, release blocks, remove from `running` (synchronized with scheduler internal list).
3. If this step determines **request ended**: `request_id` added to **`_pending_free_ids`**, and in the **next step** Worker command, notify subprocess **`free_allocation`** (avoids conflict with this step's KV usage).
4. Construct **`TokenOutput`** (`token_id`, full `text`, `finished`, `finish_reason`), **`ctx.queue.put_nowait`**, wakes `await ctx.queue.get()` in `add_request`.

---

## 9. End to cleanup

- **Normal end**: Streaming/non-streaming client reads generator to completion; `add_request`'s `finally` removes `_request_contexts` entry.
- **`abort_request`**: Scheduler aborts + pushes `FINISHED_ABORTED` to queue.
- **Process exit**: `stop()` sends shutdown, Worker exits, `join`/`terminate`.

---

## 10. In-process mode in tests and benchmarks

`AsyncLLMEngine(..., in_process=True)` uses **threads + `queue.Queue`** to run `WorkerProcess` in the same process, convenient for no-NPU or unit tests; logic is consistent with multiprocess path, only queue and startup method differ.

---

## 11. Key Configuration Reference

| Config item (Async ServingConfig) | Meaning |
|-----------------------------------|---------|
| `max_num_running_reqs` | Upper limit of concurrent running requests |
| `max_num_scheduled_tokens` | Per-scheduling-round token total budget (batch size) |
| `long_prefill_token_threshold` | Max tokens per single prefill (chunk size) |
| `max_seq_len` | Sequence length upper limit (related to block count estimation) |
| `block_size` | Aligned with KV page / `BlockPool` block size, usually equals `runtime.page_size` |
| `engine_loop_interval` | Main loop sleep interval when idle |

---

## 12. Source Code Index (for reference reading)

| Module | Path |
|--------|------|
| CLI entry and `run_serve` | `python/cli/main.py` |
| Async engine and engine loop | `python/core/async_engine.py` |
| HTTP and SSE | `python/core/server.py` |
| Scheduling and preemption | `python/core/scheduler.py` |
| Worker and subprocess entry | `python/core/worker.py` |
| Queue message structures | `python/core/types.py` (`WorkerCommand`, `StepOutput`) |
| E2E tests | `tests/test_serving_e2e.py` |
| Performance test script | `tests/bench_serving.py` |

---

## 13. Dependencies

- Serving mode requires **`fastapi`, `uvicorn`, `pydantic`**; streaming depends on SSE response format consistency (code directly writes `data: ...` lines).
- If dependencies are missing, `server.py` / `run_serve` throws clear hints at import time.

The above is a complete step-by-step description from **startup → HTTP → enqueue scheduling → Worker execution → return decode → SSE/JSON response**, which can be cross-referenced with source code section by section.
